Adding support for CAMELS IND in the datasetzoo

saurabh-3110 opened the pull request:
I have added support for the CAMELS-IND dataset following the structure and conventions used for other CAMELS datasets in the repo.

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,345 @@
+import os
+import pandas as pd
+from pathlib import Path
+from tqdm import tqdm
+from functools import reduce
+from typing import List, Dict, Union
+import xarray
+from neuralhydrology.datasetzoo.basedataset import BaseDataset
+from neuralhydrology.utils.config import Config
```
The inputs are not correctly sorted and in general it seems like the file was not formatted. Would be great if you could run yapf over this file. If you have the default neuralhydrology environment installed, you can either modify you IDE to autoformat on save (e.g. possible in visual studio vode) and by pointing to the .style.yapf file. Or you run yapf on the file manually from the terminal with something like

```
/path/to/yapf -i /path/to/camelsind.py --style /path/to/.style.yapf
```

the path to the yapf binary can be found by `which yapf` if you have the conda environment activated. The .style.yapf file is in the root directory of this repository.

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,345 @@
+
+
+def _process_and_merge_attributes(base_path: Path, attributes_output_file: Path):
+    """
+    Handles the merging of all static attribute files into a single CSV.
+    """
```
Same as above

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,345 @@
+
+# --- DOCSTRING UPDATED ---
+def preprocess_camels_ind_dataset(camels_base_path: str, output_dir: str):
+    """
+    Orchestrates the full preprocessing of a CAMELS-IND dataset. The dataset can be downloaded from
+    <https://zenodo.org/records/14999580>
```
Missing one line doc string. Also I think the link will not correctly render in the sphinx docs. Maybe use the way you also reference the dataset in the doc string of the class?

kratzert reviewed (commented):
Hi, thanks for opening this CL and for adding support for the CAMELS-IND dataset. Besides the line by line comments, a few additional points:

- for multi-basin neural network training, you usually want to work with mm/day discharge and not cms. I am not sure how the data is provided in CAMELS-IND but a few of our other classes (e.g. check CAMELSUS) have a conversion from cms to mmd implemented. You might want to check if that is needed or not.
- You haven't added your class to the docs, i.e. they would not appear on our online documentation. Can you make sure to add the class to docs/source/api ? Just check any of the other dataset classes there for what to do. It will be mostly copy and pasting. Besides creating a `neuralhydrology.datasetzoo.camelsind.rst` file, you also want to add https://github.com/neuralhydrology/neuralhydrology/blob/master/docs/source/api/neuralhydrology.datasetzoo.rst
- Please run the yapf code formatter over the file (as stated in one comment)

Also: I haven't checked the conversion code in detail and trust you that it works. The functions are rather long but I currently don't have much time to see if they could be better structure. Also probably not the most important thing.

saurabh-3110 commented:
@kratzert thanks for reviewing my code. I have made changes to the code according to your comments.
1. The code now converts the streamflow to mm/day which was earlier m3/s
2. Modified the docs
3. Correctly formatted the doc-strings for the functions and corrected the other formatting errors
4. Hardcoded the paths wherever necessary
5. Converted the loading functions public 
6. Added a check to make sure that if cfg.data_dir is in correct format or needs preprocessing

I hope these changes are enough. Let me know if any more changes are required.

kratzert commented on docs/source/api/neuralhydrology.datasetzoo.camelsind.rst:
```diff
@@ -0,0 +1,7 @@
+CamelsIND
+========
```
I see the docs test failing. I assume this is because the underlining of the title has to be of the exact same lengths as the title. If I see correctly, there is one "=" missing.
```suggestion
CamelsIND
=========
```

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,410 @@
+    References
+    ----------
+    .. [#] Mangukiya, N. K., Kumar, K. B., Dey, P., Sharma, S., Bejagam, V., Mujumdar, P. P., & Sharma, A. (2025). 
+    CAMELS-IND: hydrometeorological time series and catchment attributes for 228 catchments in Peninsular India. 
+    Earth System Science Data, 17(2), 461-491.
```
I think these two lines need to be intended to the same level as e.g. the explanation of the input Parameters.
```suggestion
        CAMELS-IND: hydrometeorological time series and catchment attributes for 228 catchments in Peninsular India. 
        Earth System Science Data, 17(2), 461-491.
```

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,410 @@
+    timeseries_output_dir : Path
+        The path to the directory where the processed, per-basin time series
+        CSV files will be saved. It will be created if it doesn't exist.
+
+    Returns
+    -------
+    None
+        This function does not return a value but saves files to disk and
+        prints summary information to the console.
```
I think the entire returns section needs to be removed.

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,410 @@
+    attributes_output_file : Path
+        The full path, including filename, where the final merged attributes
+        CSV file will be saved.
+
+    Returns
+    -------
+    None
+        This function does not return a value but saves a file to disk and
+        prints progress and summary information to the console.
```
I think the entire returns section needs to be removed.

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,410 @@
+    print(f"Consolidated attributes file saved to: {attributes_output_file}")
+
+
+# --- DOCSTRING UPDATED ---
```
Remove

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,410 @@
+    
+    The raw dataset can be downloaded from `Zenodo <https://zenodo.org/records/14999580>`_.
+    The output_dir is the processed CAMELS IND directory mentioned in the codes above.
+    This function performs two main tasks:
+    1.  Processes time series data (forcings, streamflow) and saves one CSV
```
You might need to have a blank line before this enumeration here, for the list to correctly render.

kratzert commented on neuralhydrology/datasetzoo/camelsind.py:
```diff
@@ -0,0 +1,410 @@
+
+    The final folder structure will look like this:
+    
+    output_dir/
+    ├── attributes.csv
+    └── preprocessed/
+        ├── [gauge_id_1].csv
+        ├── [gauge_id_2].csv
+        └── ...
```
I am not sure if this works. Sphinx (the docs engine) is complaining about unexpected indentation. You might need to wrap this somehow, maybe as a code block.

kratzert reviewed (commented):
There still are a few issues with the docs strings that prevent the Sphinx docs from being built. See Line 99-102 in this test https://github.com/neuralhydrology/neuralhydrology/actions/runs/16467592880/job/46548803896?pr=250

I made a guess in my line-by-line comments what could be the issue, but I am not 100% sure. To iterate more quickly, I suggest you locally install the doc dependencies from https://github.com/neuralhydrology/neuralhydrology/blob/master/environments/rtd_requirements.txt so that you can try building the docs locally to make sure they work.

If you have a neuralhydrology environment, the only things missing are:

- sphinx>=3.2.1
- sphinx-rtd-theme>=0.5.0
- nbsphinx>=0.8.0
- nbsphinx-link>=1.3.0

Once they are installed, and assuming you have a linux environment, you can go into the docs/ directory and run `make html`. You will then see a `build/` directory with an `html/` subdirectory. If the `make html` build process finishes without errors, we should be good to go. If you want to inspect the docs locally, just open the `index.html` inside the `build/html` directory, once they finish bulding.

Let me know if the instructions are unclear or if I can help you with anything.

saurabh-3110 commented:
@kratzert Thanks for taking the time to review and helping me out. I have made the necessary changes and the code should now pass the docs check and be good to go.

kratzert reviewed (approved):
LGTM
